sql/src/main/java/org/apache/druid/sql/AbstractStatement.java

```diff
@@ -0,0 +1,203 @@
+  {
+    return queryPlus.context().getAsString(PlannerContext.CTX_SQL_QUERY_ID);
+  }
+
+  /**
+   * Validate SQL query and authorize against any datasources or views which
+   * will take part in the query. Must be called by the API methods, not
+   * directly.
+   */
+  protected void validate(DruidPlanner planner)
+  {
+    plannerContext = planner.getPlannerContext();
+    plannerContext.setAuthenticationResult(queryPlus.authResult());
+    plannerContext.setParameters(queryPlus.parameters());
+    try {
+      planner.validate();
+    }
+    // We can't collapse catch clauses since SqlPlanningException has
+    // type-sensitive constructors.
+    catch (SqlParseException e) {
+      throw new SqlPlanningException(e);
+    }
+    catch (ValidationException e) {
+      throw new SqlPlanningException(e);
+    }
+  }
+
+  /**
+   * Authorize the query using the authorizer provided, and an option to authorize
+   * context variables as well as query resources.
+   */
+  protected void authorize(
+      DruidPlanner planner,
+      Function<Set<ResourceAction>, Access> authorizer
+  )
+  {
+    boolean authorizeContextParams = sqlToolbox.authConfig.authorizeQueryContextParams();
+
+    // Authentication is done by the planner using the function provided
+    // here. The planner ensures that this step is done before planning.
+    Access authorizationResult = planner.authorize(authorizer, authorizeContextParams);
+    if (!authorizationResult.isAllowed()) {
+      throw new ForbiddenException(authorizationResult.toString());
+    }
+
+    queryResource = planner.resourceActions(false);
+    resourceActions = planner.resourceActions(authorizeContextParams);
+  }
+
+  /**
+   * Resource authorizer based on the authentication result
+   * provided earlier.
+   */
+  protected Function<Set<ResourceAction>, Access> authorizer()
+  {
+    return resourceActions ->
+      AuthorizationUtils.authorizeAllResourceActions(
+          queryPlus.authResult(),
+          resourceActions,
+          sqlToolbox.plannerFactory.getAuthorizerMapper()
+    );
+  }
+
+  /**
+   * Plan the query, which also produces the sequence that runs
+   * the query.
+   */
+  protected PlannerResult plan(DruidPlanner planner)
+  {
+    try {
+      return planner.plan();
+    }
+    catch (ValidationException e) {
+      throw new SqlPlanningException(e);
+    }
+  }
+
+  /**
+   * Return the datasource and table resources for this
+   * statement.
+   */
+  public Set<ResourceAction> resources()
+  {
+    return queryResource;
+  }
+
+  public Set<ResourceAction> allResources()
+  {
+    return resourceActions;
+  }
+
+  public SqlQueryPlus sqlRequest()
+  {
+    return queryPlus;
+  }
+
+  public SqlExecutionReporter reporter()
+  {
+    return reporter;
+  }
+
+  /**
+   * Releases resources and emits logs and metrics as defined the
+   * associated reporter.
+   */
+  @Override
+  public void close()
+  {
+    closeQuietly();
+    reporter.emit();
```

Comment:
I'm wondering if there should be a try/finally here. Can `closeQuietly` throw errors? If so, should we still call `reporter.emit()` in that case?